packages/SwingSet/src/vats/vat-admin/vat-vat-admin.js

```diff
@@ -41,88 +169,114 @@ export function buildRootObject(vatPowers) {
-    const meter = Far('meter', {
-      addRemaining(_delta) {},
-      setThreshold(_newThreshold) {},
-      get: () => harden({ remaining: 'unlimited', threshold: 0 }),
-      getNotifier: () => notifier, // will never fire
-    });
-    meterByID.set(meterID, harden({ meter, updater }));
-    meterIDByMeter.set(meter, meterID);
-    return meter;
+  function resumeService() {
+    if (quiescentP) {
+      quiescentRR.reject(Error('shutdown cancelled'));
+      quiescentP = null;
+      quiescentRR = null;
+    }
   }
 
-  function finishVatCreation(vatID) {
-    const [promise, pendingRR] = producePRR();
-    pending.set(vatID, pendingRR);
+  function finishMeter({ state, self }) {
+    meterByID.init(
+      state.meterID,
+      // harden({ meter: self, updater: state.updater }), // XXX RESTORE
+      harden({ meter: self }), // XXX TEMP
+    );
+    meterIDByMeter.set(self, state.meterID);
+    meterNotifiersTemp.set(self, makeNotifierKit()); // XXX TEMP
+  }
 
-    const [doneP, doneRR] = producePRR();
-    running.set(vatID, doneRR);
-    doneP.catch(() => {}); // shut up false whine about unhandled rejection
+  function insistServiceAvailable() {
+    assert(!quiescentP, 'service not available, try again later');
+  }
+
+  async function upgradeVat(vatID, readyP, bundleID, options = {}) {
+    assert(vatAdminDev, `vatAdmin device not configured`);
+    const { vatParameters, upgradeMessage = 'vat upgraded', ...rest } = options;
+    const leftovers = Object.keys(rest);
+    if (leftovers.length) {
+      const bad = leftovers.join(',');
+      assert.fail(X`upgrade() received unknown options: ${bad}`);
+    }
+    assert.typeof(upgradeMessage, 'string', 'upgradeMessage not a string');
+    if (readyP) {
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      await readyP.catch(() => true);
+    }
+    if (vatID === vatAdminVatID) {
+      baggage.init('doingSelfUpgrade', true);
+    }
+    const upgradeID = D(vatAdminDev).upgradeVat(
+      vatID,
+      bundleID,
+      vatParameters,
+      upgradeMessage,
+    );
+    const [upgradeCompleteP, upgradeRR] = producePRR();
+    pendingUpgrades.set(upgradeID, upgradeRR);
+    return upgradeCompleteP;
+  }
+
+  function assertRunningVat(vatID, operation, exitingOK) {
+    assert(
+      runningVats.has(vatID) && (exitingOK || !exitingVats.has(vatID)),
+      `attempt to perform ${operation}() on non-running vat ${vatID}`,
+    );
+  }
 
-    const adminNode = Far('adminNode', {
-      terminateWithFailure(reason) {
-        D(vatAdminNode).terminateWithFailure(vatID, reason);
+  const adminNodeKindHandle = provide(baggage, 'adminNode', () =>
+    makeKindHandle('adminNode'),
+  );
+  const makeAdminNode = defineDurableKind(
+    adminNodeKindHandle,
+    vatID => ({ vatID }),
+    {
+      terminateWithFailure({ state }, reason) {
+        insistServiceAvailable();
+        assertRunningVat(state.vatID, 'terminateWithFailure');
+        exitingVats.add(state.vatID);
```

Comment:
I'm trying to analyze this bit. So when someone calls `adminFacet~.terminateWithFailure()` for some random dynamic vat, we'll populate `exitingVats`, which will inhibit vat-vat-admin upgrade until the dynamic vat actually terminates (and we find out when the `vatAdminRoot.vatTerminated` callback gets run, and we do `exitingVats.delete` and then re-check for quiescence). The `assetRunningVat` protects against the `terminateWithFailure` arriving *after* the vat terminated (perhaps spontaneously, perhaps because of a rival+winning `terminateWithFailure` call): without it, we might populate `exitingVats` but have no way to un-populate it, which would inhibit future self-upgrades (although we could hack our way around that, if we needed to, by doing a chain upgrade with `controller.upgradeStaticVat('vatAdmin', shouldPauseFirst=false)`).

There's a small behavior change here relative to the old version, which seems to believe that it tolerates multiple (or just very very late) `terminateWithFailure` calls. I don't know what `D(vatAdminDev).terminateWithFailure` does in that case.. I suspect it wasn't as tolerant as the code made it appear.

So let's see, if someone fetches a vat's `done` promise with `adminFacet~.done()` before the vat gets terminated and before vat-vat-admin starts a self-upgrade, that promise will inhibit vat-vat-admin self-upgrade until the vat terminates. Which means that `done` promise *won't* reject because of vat-vat-admin self-upgrade. It will either fulfill (vat terminate-happy), or reject (vat terminate-sad). So they can rely upon rejection being a real vat-terminate-sad event.

If their `done()` request arrives after the vat is terminated, it rejects from the `assertRunningVat` message.

If it arrives after vat-vat-admin has been paused, it rejects from the `insistServiceAvailable()` check.

So if the client calls `done()` and gets a rejection, it could be any of:
* vat-terminate-sad
* vat was already terminated, we don't know its state
* vat-vat-admin is paused

and they can't easily tell the difference.